File: mip_reference/ruby-sinatra/lib/mip/store.rb
# frozen_string_literal: true

module MIP
  # In-memory data store for MIP node data
  # Uses a simple singleton pattern for this reference implementation
  class Store
    class << self
      attr_accessor :instance
    end

    attr_reader :node_identity, :connections, :members, :endorsements,
                :search_requests, :cogs_requests, :activity_log

    def initialize
      @node_identity = nil
      @connections = {}      # mip_identifier => Connection
      @members = {}          # member_number => Member
      @endorsements = {}     # endorsement_id => Endorsement
      @search_requests = {}  # shared_identifier => SearchRequest
      @cogs_requests = {}    # shared_identifier => CogsRequest
      @activity_log = []     # Array of activity entries
    end

    def set_node_identity(identity)
      @node_identity = identity
    end

    # Connection management
    def add_connection(connection)
      @connections[connection.mip_identifier] = connection
      log_activity("Connection added: #{connection.organization_name} (#{connection.status})")
    end

    def find_connection(mip_identifier)
      @connections[mip_identifier]
    end

    def active_connections
      @connections.values.select(&:active?)
    end

    def pending_connections
      @connections.values.select(&:pending?)
    end

    def all_connections
      @connections.values
    end

    # Member management
    def add_member(member)
      @members[member.member_number] = member
    end

    def find_member(member_number)
      @members[member_number]
    end

    def search_members(query)
      @members.values.select do |m|
        if query[:member_number]
          m.member_number.downcase.include?(query[:member_number].downcase)
        elsif query[:first_name] && query[:last_name]
          name_match = m.first_name.downcase.include?(query[:first_name].downcase) &&
                       m.last_name.downcase.include?(query[:last_name].downcase)
          if query[:birthdate]
            name_match && m.birthdate == query[:birthdate]
          else
            name_match
          end
        else
          false
        end
      end
    end

    def all_members
      @members.values
    end

    # Endorsement management
    def add_endorsement(endorsement)
      @endorsements[endorsement.id] = endorsement
      log_activity("Endorsement received from #{endorsement.endorser_mip_identifier}")
    end

    def find_endorsements_for(mip_identifier)
      @endorsements.values.select { |e| e.endorsed_mip_identifier == mip_identifier }
    end

    def find_endorsements_from(mip_identifier)
      @endorsements.values.select { |e| e.endorser_mip_identifier == mip_identifier }
    end

    # Search request management
    def add_search_request(search_request)
      @search_requests[search_request.shared_identifier] = search_request
      log_activity("Search request: #{search_request.direction} - #{search_request.target_org}")
    end

    def find_search_request(shared_identifier)
      @search_requests[shared_identifier]
    end

    def all_search_requests
      @search_requests.values
    end

    # COGS request management
    def add_cogs_request(cogs_request)
      @cogs_requests[cogs_request.shared_identifier] = cogs_request
      log_activity("COGS request: #{cogs_request.direction} - #{cogs_request.target_org}")
    end

    def find_cogs_request(shared_identifier)
      @cogs_requests[shared_identifier]
    end

    def all_cogs_requests
      @cogs_requests.values
    end

    # Activity log
    def log_activity(message)
      @activity_log.unshift({
        timestamp: Time.now.iso8601,
        message: message
      })
      # Keep only last 100 entries
      @activity_log = @activity_log.first(100)
    end

    def recent_activity(count = 20)
      @activity_log.first(count)
    end

    # Class method for global access
    def self.current
      @instance ||= new
    end

    def self.reset!
      @instance = new
    end
  end
end

File: mip_reference/node-express/lib/mip/store.js
'use strict';

// In-memory data store for MIP node data
// Uses a simple singleton pattern for this reference implementation
class Store {
  constructor() {
    this.nodeIdentity = null;
    this.connections = new Map();       // mip_identifier => Connection
    this.members = new Map();           // member_number => Member
    this.endorsements = new Map();      // endorsement_id => Endorsement
    this.searchRequests = new Map();    // shared_identifier => SearchRequest
    this.cogsRequests = new Map();      // shared_identifier => CogsRequest
    this.activityLog = [];              // Array of activity entries
  }

  setNodeIdentity(identity) {
    this.nodeIdentity = identity;
  }

  // Connection management
  addConnection(connection) {
    this.connections.set(connection.mipIdentifier, connection);
    this.logActivity(`Connection added: ${connection.organizationName} (${connection.status})`);
  }

  findConnection(mipIdentifier) {
    return this.connections.get(mipIdentifier) || null;
  }

  activeConnections() {
    return [...this.connections.values()].filter(c => c.isActive());
  }

  pendingConnections() {
    return [...this.connections.values()].filter(c => c.isPending());
  }

  allConnections() {
    return [...this.connections.values()];
  }

  // Member management
  addMember(member) {
    this.members.set(member.memberNumber, member);
  }

  findMember(memberNumber) {
    return this.members.get(memberNumber) || null;
  }

  searchMembers(query) {
    return [...this.members.values()].filter(m => {
      if (query.member_number) {
        return m.memberNumber.toLowerCase().includes(query.member_number.toLowerCase());
      } else if (query.first_name && query.last_name) {
        const nameMatch = m.firstName.toLowerCase().includes(query.first_name.toLowerCase()) &&
                          m.lastName.toLowerCase().includes(query.last_name.toLowerCase());
        if (query.birthdate) {
          return nameMatch && m.birthdate === query.birthdate;
        }
        return nameMatch;
      }
      return false;
    });
  }

  allMembers() {
    return [...this.members.values()];
  }

  // Endorsement management
  addEndorsement(endorsement) {
    this.endorsements.set(endorsement.id, endorsement);
    this.logActivity(`Endorsement received from ${endorsement.endorserMipIdentifier}`);
  }

  findEndorsementsFor(mipIdentifier) {
    return [...this.endorsements.values()].filter(e => e.endorsedMipIdentifier === mipIdentifier);
  }

  findEndorsementsFrom(mipIdentifier) {
    return [...this.endorsements.values()].filter(e => e.endorserMipIdentifier === mipIdentifier);
  }

  // Search request management
  addSearchRequest(searchRequest) {
    this.searchRequests.set(searchRequest.sharedIdentifier, searchRequest);
    this.logActivity(`Search request: ${searchRequest.direction} - ${searchRequest.targetOrg}`);
  }

  findSearchRequest(sharedIdentifier) {
    return this.searchRequests.get(sharedIdentifier) || null;
  }

  allSearchRequests() {
    return [...this.searchRequests.values()];
  }

  // COGS request management
  addCogsRequest(cogsRequest) {
    this.cogsRequests.set(cogsRequest.sharedIdentifier, cogsRequest);
    this.logActivity(`COGS request: ${cogsRequest.direction} - ${cogsRequest.targetOrg}`);
  }

  findCogsRequest(sharedIdentifier) {
    return this.cogsRequests.get(sharedIdentifier) || null;
  }

  allCogsRequests() {
    return [...this.cogsRequests.values()];
  }

  // Activity log
  logActivity(message) {
    this.activityLog.unshift({
      timestamp: new Date().toISOString(),
      message
    });
    // Keep only last 100 entries
    if (this.activityLog.length > 100) {
      this.activityLog = this.activityLog.slice(0, 100);
    }
  }

  recentActivity(count = 20) {
    return this.activityLog.slice(0, count);
  }
}

// Singleton
let instance = null;

function current() {
  if (!instance) instance = new Store();
  return instance;
}

function reset() {
  instance = new Store();
  return instance;
}

module.exports = { Store, current, reset };
